Stm32c0 add power mgmt

jurenat opened the pull request:
Adds support for STOP mode. It is based on the F4 code and uses the RTC as an idle timer. From my tests with NUCLEO-C071RB, the power consumption goes down from ~3mA to 90uA. Tested with the samples/basic/blinky and samples/boards/st/power_mgmt/blinky/

jurenat committed:
soc: st: stm32: stm32c0x: Add stop mode support

Enables low power stop mode for C0.

Code is taken from F4 family, tested on nucleo-c71rb with
samples/basic/blinky.

Power consumption in run mode 3.7 mA, in stop mode ~87 uA.

Signed-off-by: Tomáš Juřena <jurenatomas@gmail.com>

jurenat committed:
dts: arm: st: c0: Add power states

Adds power-states node to support stop mode. As idle timer is used RTC.

Signed-off-by: Tomáš Juřena <jurenatomas@gmail.com>

erwango reviewed (changes requested):
Thanks for adding this.

Also, it would be nice that we have a C0 board being tested with `samples/boards/st/power_mgmt/blinky`. Can you make sure that at least one board is compatible with this sample and that it is not filtered out by twister ?

dsseng commented on soc/st/stm32/stm32c0x/power.c:
```diff
@@ -0,0 +1,77 @@
+		stm32_clock_control_init(NULL);
+		break;
+	default:
+		LOG_DBG("Unsupported power substate-id %u", state);
```
This seems to fall into the `programmer error` class

```suggestion
		LOG_WRN("Unsupported power substate-id %u", state);
```

jurenat committed:
samples: boards: st: power_mgmt: blinky: Enable NUCLEO-C071RB

Enables the NUCLEO-C071RB board to the blinky sample to be run by twister.

Signed-off-by: Tomáš Juřena <jurenatomas@gmail.com>

jurenat commented on soc/st/stm32/stm32c0x/power.c:
```diff
@@ -0,0 +1,77 @@
+
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/interrupt_controller/gic.h>
```
Removed

jurenat commented on soc/st/stm32/stm32c0x/power.c:
```diff
@@ -0,0 +1,77 @@
+#include <stm32c0xx.h>
+
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/counter.h>
```
Removed

jurenat commented on soc/st/stm32/stm32c0x/power.c:
```diff
@@ -0,0 +1,77 @@
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/pm.h>
+#include <zephyr/init.h>
```
```
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
```
are needed, rest removed

jurenat commented on soc/st/stm32/stm32c0x/soc.c:
```diff
@@ -32,4 +34,7 @@ void soc_early_init_hook(void)
 	/* Update CMSIS SystemCoreClock variable (HCLK) */
 	/* At reset, system core clock is set to 48 MHz from HSI */
 	SystemCoreClock = 48000000;
+#if CONFIG_PM
```
Replaced by call in `soc_early_init_hook`

jurenat commented on soc/st/stm32/stm32c0x/power.c:
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2023 Google LLC
```
Updated

jurenat committed:
drivers: counter: stm32: Disable BYPSHAD in init

Disable the shadow registers bypass in init function before RTC
deinitialisation to avoid infinite loop as the bit cases the RSF bit
to be kept in 0. The systick is not working in PRE_KERNEL_1 so it will
never timeout.

Signed-off-by: Tomáš Juřena <jurenatomas@gmail.com>
